Add OutputType::BECH32M and related wallet support for fetching bech32m addresses

achow101 opened the pull request:
Currently bech32m addresses are classfied as bech32. Because bech32m is incompatible with bech32, we need to define a new `OutputType` for it so that it can be handled correctly. This PR adds `OutputType::BECH32M`, updates all of the relevant `OutputType` classifications, and handle requests for bech32m addresses. There is now a `bech32m` address type string that can be used.

* `tr()` descriptors now report their output type as `OutputType::BECH32M`. `WtinessV1Taproot` and `WitnessUnknown` are also classified as `OutputType::BECH32M`.
* Bech32m addresses are completely disabled for legacy wallets. They cannot be imported (explicitly disallowed in `importaddress` and `importmulti`), will not be created when getting all destinations for a pubkey, and will not be added with `addmultisigaddress`. Additional protections have been added to `LegacyScriptPubKeyMan` to disallow attempting to retrieve bech32m addresses.
* Since Taproot multisigs are not implemented yet, `createmultisig` will also disallow the bech32m address type.
* As Taproot is not yet active, `DescriptorScriptPubKeyMan` cannot and will not create a `tr()` descriptor. Protections have been added to make sure this cannot occur.
* The change address type detection algorithm has been updated to return `bech32m` when there is a segwit v1+ output script and the wallet has a bech32m `ScriptPubKeyMan`, falling back to bech32 if one is not available.

achow101 committed:
Limit LegacyScriptPubKeyMan address types

Make sure that LegacyScriptPubKeyMan can only be used for legacy,
p2sh-segwit, and bech32 address types.

achow101 committed:
Add OutputType::BECH32M

Bech32m addresses need their own OutputType

We are not ready to create DescriptorScriptPubKeyMans which produce
bech32m addresses. So don't allow generating them.

achow101 committed:
Opportunistically use bech32m change addresses if available

If a transaction as a segwit output, use a bech32m change address if
they are available. If not, fallback to bech32. If bech32 change
addresses are unavailable, fallback to the default address type.

achow101 committed:
Use BECH32M for tr() desc, WitV1Taproot, and WitUnknown CTxDests

The tr() descriptor, WitnessV1Taproot CTxDestination, and
WitnessUnknown CTxDestination are OutputType::BECH32M so they should
report as such.

achow101 committed:
Disallow bech32m addresses for legacy wallet things

We don't want the legacy wallet to ever have bech32m addresses so don't
allow importing them. This includes addmultisigaddress as that is a
legacy wallet only RPC

Additionally, bech32m multisigs are not available yet, so disallow them
in createmultisig.

sipa commented on src/rpc/misc.cpp:
```diff
@@ -131,6 +131,9 @@ static RPCHelpMan createmultisig()
             throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("Unknown address type '%s'", request.params[2].get_str()));
         }
+        if (output_type == OutputType::BECH32M) {
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Bech32m multisig addresses cannot be created yet");
```
Do we intend to ever create those through this RPC? I think you'd use `deriveaddress` or so instead.

achow101 replied:
Probably not. Changed the message.

fjahr commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -23,6 +23,11 @@ const uint32_t BIP32_HARDENED_KEY_LIMIT = 0x80000000;
 
 bool LegacyScriptPubKeyMan::GetNewDestination(const OutputType type, CTxDestination& dest, std::string& error)
 {
+    if (LEGACY_OUTPUT_TYPES.count(type) == 0) {
```
Have you thought about placing asserts in `LegacyScriptPubKeyMan::KeepDestination`, `LegacyScriptPubKeyMan::GetKeyFromPool`, and `LegacyScriptPubKeyMan::LearnRelatedScripts`?

achow101 replied:
Added asserts for all of the `LegacyScriptPubKeyMan` functions which take an `OutputType`.

fjahr commented on src/wallet/wallet.cpp:
```diff
@@ -1906,7 +1906,12 @@ OutputType CWallet::TransactionChangeType(const std::optional<OutputType>& chang
         int witnessversion = 0;
         std::vector<unsigned char> witnessprogram;
-        if (recipient.scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram)) {
+        bool is_wit = recipient.scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram);
```
Can be `const`

achow101 replied:
Done

fjahr reviewed (commented):
Looks close, only some style nits and open questions/suggestions.

achow101 commented on src/wallet/scriptpubkeyman.h:
```diff
@@ -254,6 +254,13 @@ class ScriptPubKeyMan
 };
 
+/** OutputTypes supported by the LegacyScriptPubKeyMan */
+static const std::unordered_set<OutputType> LEGACY_OUTPUT_TYPES {
```
Tried that but it doesn't work.

sipa replied:
You can't `constexpr` things that need allocation. You could use an array instead of an unordered_set, and use `std::find` to look up (which will likely be faster too, due to the tiny size), but it really doesn't matter here.

achow101 committed:
wallet: Add error message to GetReservedDestination

Adds an error output parameter to all GetReservedDestination functions
so that callers can get the actual reason that a change address could
not be fetched. This more closely matches GetNewDestination. This allows
for more granular error messages, such as one that indicates that
bech32m addresses cannot be generated yet.

achow101 commented on test/functional/wallet_address_types.py:
```diff
@@ -373,5 +373,15 @@ def run_test(self):
+            self.log.info("Descriptor wallets do not have bech32m addreses by default yet")
+            # TODO: Remove this when they do
+            assert_raises_rpc_error(-12, "Error: No bech32m addresses available", self.nodes[0].getnewaddress, "", "bech32m")
+            assert_raises_rpc_error(-12, "Error: Keypool ran out, please call keypoolrefill first", self.nodes[0].getrawchangeaddress, "bech32m")
```
I've added a commit which allows `GetReservedDestination` to return real error messages. So this is now `No bech32m addresses available`.

achow101 commented on test/functional/wallet_address_types.py:
```diff
@@ -373,5 +373,15 @@ def run_test(self):
         self.test_address(4, self.nodes[4].getrawchangeaddress('bech32'), multisig=False, typ='bech32')
 
+        if self.options.descriptors:
+            self.log.info("Descriptor wallets do not have bech32m addreses by default yet")
```
Fixed

fjahr commented:
utACK 05ff76c3aad6905d984ba1764c8045ca530fbf8f

Sjors reviewed (approved):
tACK 05ff76c

I made a followup PR that makes the defaults a bit more friendly, and adds GUI support: #22260

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -23,6 +23,11 @@ const uint32_t BIP32_HARDENED_KEY_LIMIT = 0x80000000;
 bool LegacyScriptPubKeyMan::GetNewDestination(const OutputType type, CTxDestination& dest, std::string& error)
 {
+    if (LEGACY_OUTPUT_TYPES.count(type) == 0) {
+        error = _("Error: Legacy wallets only support the \"legacy\", \"p2sh-segwit\", and \"bech32\" address types").translated;
```
Yes, just belt and suspenders.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1906,7 +1906,12 @@ OutputType CWallet::TransactionChangeType(const std::optional<OutputType>& chang
         std::vector<unsigned char> witnessprogram;
-        if (recipient.scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram)) {
+        const bool is_wit = recipient.scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram);
+        if (is_wit && witnessversion >= 1 && GetScriptPubKeyMan(OutputType::BECH32M, true)) {
```
Since this is generic for future segwit versions, I'm not sure doing that makes sense. For taproot specifically, there won't be a bech32m `ScriptPubKeyMan` until taproot activates anyways.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -1906,7 +1906,12 @@ OutputType CWallet::TransactionChangeType(const std::optional<OutputType>& chang
+            // If this is a segwit v1+ address and this wallet supports bech32m addresses, use bech32m
+            return OutputType::BECH32M;
+        } else if (is_wit) {
+            // Otherwise if this is witness, use bech32
```
This assumes there is always a bech32 descriptor, which may not be the case if you're using a custom taproot descriptor without fallback.

```patch
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 3eca506f08..c0b28acd47 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1911,9 +1911,11 @@ OutputType CWallet::TransactionChangeType(const std::optional<OutputType>& chang
         if (is_wit && witnessversion >= 1 && GetScriptPubKeyMan(OutputType::BECH32M, true)) {
             // If this is a segwit v1+ address and this wallet supports bech32m addresses, use bech32m
             return OutputType::BECH32M;
-        } else if (is_wit) {
-            // Otherwise if this is witness, use bech32
+        } else if (is_wit && GetScriptPubKeyMan(OutputType::BECH32, true)) {
+            // Otherwise if this is witness, use bech32 if we have a descriptor:
             return OutputType::BECH32;
+        } else {
+            return m_default_address_type;
         }
     }
```

achow101 replied:
The original code did not check for a bech32 `ScriptPubKeyMan`, but I have added that check. I did not include the `else` suggested in your diff as that would be incorrect.

Sjors commented:
re-utACK f70d55c136d9fc0bc986cc86e619570e0417d188

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -1906,7 +1906,12 @@ OutputType CWallet::TransactionChangeType(const std::optional<OutputType>& chang
         std::vector<unsigned char> witnessprogram;
-        if (recipient.scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram)) {
+        const bool is_wit = recipient.scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram);
+        if (is_wit && witnessversion >= 1 && GetScriptPubKeyMan(OutputType::BECH32M, true)) {
```
4bf448fc2528d9df371b2ca6e9b13319441a21e5: I don't understand what the `witnessversion >= 1` is doing.

Assuming `witnessversion` is `2`, then a bech32m address will be selected (possibly a taproot address). However, if the `witnessversion` is `0`, even with a taproot wallet, a segwit_0 address will be enforced. This seems inconsistent and confusing, unless I am missing something?

Sjors replied:
This opportunistic change code was initially intended to _upgrade_ change addresses to `bech32` if there was no privacy downside. With this new change it seems that we _downgrade_ a v1+ witness to v0 just in order to blend in. I'm also not sure if we should do that. It could actually hurt privacy when that v0 change address is spent later on and combined with a v1 input.

achow101 replied:
Indeed. I've changed this to just return `BECH32M` or `BECH32` depending on what is available, with a preference for `BECH32M`.

However I think this function (and generally how change address type is determined) should be revisited because not all address types are available in descriptor wallets. But I think that is out of scope for this PR.

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -2115,7 +2115,7 @@ bool CWallet::GetNewDestination(const OutputType type, const std::string label,
         result = spk_man->GetNewDestination(type, dest, error);
     } else {
-        error = strprintf("Error: No %s addresses available.", FormatOutputType(type));
+        error = strprintf(_("Error: No %s addresses available."), FormatOutputType(type)).translated;
```
Not sure how difficult it would be to fix, but the general idea for translations is to pass up the bilingual string and let the caller pick the right version. Otherwise the RPC responses (and debug log) might be partially translated. Functional test might fail when run with the gui.

achow101 replied:
I think that will be a bit invasive since several of the functions involved use `std::string error` rather than `bilingual_str`. Perhaps for a followup.

laanwj replied:
Yes, if you're going to translate anything, it's better to use `bilingual_str`. I would prefer to do this right in the first go, instead of having another follow-up commit to prevent RPC messages from being translated.

Alternatively just leave this untranslated for now. This is marked for 22.0 and we've passed the translation string freeze.

achow101 replied:
For the strings I added, I was just following the pattern already in use in these functions. For this change in particular, it was to make the style unified. Unless this is considered a blocker, I'm going to leave it as is.

laanwj replied:
Ok…

maflcko reviewed (commented):
left two comments.

practicalswift commented:
Concept ACK

Sjors commented:
re-utACK f33361c17f3937ee8819fcea76548e19f6ac6558

fjahr commented:
re-ACK f33361c

The only changes since my last review was a reorganization of the code changes in `TransactionChangeType()` and ensuring there that `BECH32` address types are also available before using them. That commit was also slightly renamed.

laanwj commented:
Code review ACK f33361c17f3937ee8819fcea76548e19f6ac6558

I think the potential RPC error message translation issue mentioned in https://github.com/bitcoin/bitcoin/pull/22154#discussion_r653415459 needs to be addressed at some point, but also think it's not a blocker for 22.0.

meshcollider commented:
I get an error in wallet_taproot.py:
```python
Remaining jobs: [wallet_taproot.py]
1/1 - wallet_taproot.py failed, Duration: 1 s

stdout:
2021-06-23T01:47:26.095000Z TestFramework (INFO): Initializing test directory /tmp/test_runner_₿_🏃_20210623_134725/wallet_taproot_0
2021-06-23T01:47:26.359000Z TestFramework (INFO): Creating wallets...
2021-06-23T01:47:26.605000Z TestFramework (INFO): Mining blocks...
2021-06-23T01:47:26.694000Z TestFramework (INFO): Testing tr(XPRV) address derivation
2021-06-23T01:47:26.745000Z TestFramework (INFO): Testing tr(XPRV) through sendtoaddress
2021-06-23T01:47:26.770000Z TestFramework (ERROR): JSONRPC error
Traceback (most recent call last):
  File "/home/meshcollider/bitcoin/test/functional/test_framework/test_framework.py", line 128, in main
    self.run_test()
  File "/home/meshcollider/bitcoin/test/functional/wallet_taproot.py", line 347, in run_test
    1
  File "/home/meshcollider/bitcoin/test/functional/wallet_taproot.py", line 314, in do_test
    self.do_test_sendtoaddress(comment, pattern, privmap, treefn, keys[0:nkeys], keys[nkeys:2*nkeys])
  File "/home/meshcollider/bitcoin/test/functional/wallet_taproot.py", line 262, in do_test_sendtoaddress
    addr_g = self.rpc_online.getnewaddress(address_type='bech32')
  File "/home/meshcollider/bitcoin/test/functional/test_framework/coverage.py", line 47, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/meshcollider/bitcoin/test/functional/test_framework/authproxy.py", line 146, in __call__
    raise JSONRPCException(response['error'], status)
test_framework.authproxy.JSONRPCException: Error: No bech32 addresses available. (-12)
2021-06-23T01:47:26.821000Z TestFramework (INFO): Stopping nodes
2021-06-23T01:47:27.028000Z TestFramework (WARNING): Not cleaning up dir /tmp/test_runner_₿_🏃_20210623_134725/wallet_taproot_0
2021-06-23T01:47:27.028000Z TestFramework (ERROR): Test failed. Test logging available at /tmp/test_runner_₿_🏃_20210623_134725/wallet_taproot_0/test_framework.log
```

achow101 commented:
Silent merge conflict, rebased and fixed.

Sjors commented:
re-utACK 754f134: only change is switching to `bech32m` in two `wallet_taproot.py` test cases.

jonatack commented:
Reviewing and testing.

jonatack commented on test/functional/wallet_address_types.py:
```diff
@@ -373,5 +373,15 @@ def run_test(self):
         self.test_address(4, self.nodes[4].getrawchangeaddress('bech32'), multisig=False, typ='bech32')
 
+        if self.options.descriptors:
+            self.log.info("Descriptor wallets do not have bech32m addresses by default yet")
```
s/addreses/addresses/ should be fixed one commit earlier in 87a0e7a

jonatack reviewed (commented):
ACK 754f134a50cc56cdf0baf996d909c992770fcc97

When should `bech32m` be added to the `-addresstype` and `-changetype` wallet config option helps?

achow101 commented:
> When should `bech32m` be added to the `-addresstype` and `-changetype` wallet config option helps?

When we make bech32m descriptors by default.

fjahr commented:
re-ACK 754f134a50cc56cdf0baf996d909c992770fcc97

Confirmed only changes were rebasing and fixing `wallet_taproot.py`.

Sjors commented:
I rebased my followup in #22260 (which probably works best if you don't set `-addresstype`)
